feat(service): Improve Fastly 404 error message for service permissions

marceloboeira opened the pull request:
As explained in #963

When attempting to fetch a Fastly service not included in the API key's permissions, the current error message is misleading and difficult to debug.

Before:
```
Error: 404 - Not Found:
    Title:  Record not found
    Detail: Cannot find 'nil_class'
```

After:
```
Error: Service with ID '***REDACTED***' was not found.
  This could mean the resource doesn't exist, or that the Fastly API key used doesn't have the necessary permissions.
  Please verify the resource ID and API key permissions.
```

The new error message provides clearer guidance by:
- Explicitly stating the potential reasons for the failure
- Helping developers quickly identify whether the issue is with:
  1. Service existence
  2. API key permissions

This change saves debugging time by offering more context and actionable information when encountering service-related access issues.

 All Submissions:

* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/fastly/go-fastly/pulls) for the same update/change?

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### New Feature Submissions:

* [x] Does your submission pass tests?

### Changes to Core Features:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your core changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

### User Impact

* [x] What is the user impact of this change?

### Are there any considerations that need to be addressed for release?

Should we apply this error everywhere?

<!-- Any breaking changes, etc -->

marceloboeira committed:
feat(service): Improve Fastly 404 error message for service permissions

As explained in #963

When attempting to fetch a Fastly service not included in the API key's permissions, the current error message is misleading and difficult to debug.

Before:
```
Error: 404 - Not Found:
    Title:  Record not found
    Detail: Cannot find 'nil_class'
```

After:
```
Error: Service with ID '***REDACTED***' was not found.
  This could mean the resource doesn't exist, or that the Fastly API key used doesn't have the necessary permissions.
  Please verify the resource ID and API key permissions.
```

The new error message provides clearer guidance by:
- Explicitly stating the potential reasons for the failure
- Helping developers quickly identify whether the issue is with:
  1. Service existence
  2. API key permissions

This change saves debugging time by offering more context and actionable information when encountering service-related access issues.

kpfleming commented:
Thanks for the contribution! After a quick look this appears to be very well done, including tests which we always appreciate. There have been some other changes in the source since you created your branch, so you'll need to pull the current version and resolve the conflicts.

marceloboeira commented:
@kpfleming Thanks! 

 I didn't realise my fork was so far behind. I've rebased it and fixed the incompatibilities!

My only question is: should this be fixed on other resources as well? 

Context: I'm guessing if someone faces this issue it will be with services since that's one of the main entities. However, I'm not too familiar with API key permission scopes at Fastly — so this could happen with other resources (e.g.:  tls-subscriptions, ...).

In any case the function should be re-usable for other HTTP contexts, it's just not actively done now.

kpfleming commented:
At the moment services are the only resource which can be limited on a per-token basis, so there's no immediate need to expand this.

anthony-gomez-fastly committed:
Merge branch 'main' into issue-963

anthony-gomez-fastly commented:
@marceloboeira could you please add the PR to the changelog entry?

marceloboeira commented:
@anthony-gomez-fastly it's there now :D

anthony-gomez-fastly commented:
wow i fixed the merge and it's the other one that doesn't have the PR, looks like i have something to fix 😅
